<a href="https://colab.research.google.com/github/AlinaOtr/data/blob/master/Researching_music_preferences.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div style="border: 2px solid #333; border-radius: 10px; padding: 20px; background-color: lightblue; box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);">
    <div style="display: flex; justify-content: space-around; flex-wrap: wrap;">
        <div style="text-align: left;">
            <img src="https://img.ixbt.site/live/images/original/00/06/21/2023/04/03/dddcab2e03.png" alt="Логотип проекта" style="max-width: 250px; border-radius: 10px; box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); margin-left: 0">
            <h1 style="color: #2c3e50;">Музыка больших городов</h1>        
        </div>
        <div style="display: inline-block; background-color: #f0f0f0; padding: 15px; border-radius: 10px; margin-bottom: 10px;">
            <h3 style="margin: 0; color: #2c3e50;">Гипотезы:</h3>
            <ol style="margin-top: 10px;">
                <li>Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному;</li>
                <li>Утром в понедельник в Москве преобладают одни жанры музыки, а в Петербурге — другие. Это верно и для вечера пятницы;</li>
                <li>Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.</li>
            </ol>
        </div>
    </div>
    <a name="оглавление"></a>
    <div style="background-color: #f0f0f0; padding: 10px; border-radius: 10px; margin-top: 15px;">
        <h3 style="margin: 0; color: #2c3e50;">Ход исследования:</h3>
        <ol style="margin-top: 10px;">
            <li><a href="#получение-данных">Получение данных</a></li>
            <li><a href="#предобработка-данных">Предобработка данных</a></li>
            <li><a href="#проверка-гипотез">Проверка гипотез:</a>
                <ul style="margin: 5px;">
                    <li><a href="#гипотеза1">Действительно ли музыку в разных городах слушают по-разному?</a></li>
                    <li><a href="#гипотеза2">Утро понедельника и вечер пятницы — разная музыка или одна и та же?</a></li>
                    <li><a href="#гипотеза3">Москва и Питер — две разные столицы, два разных направления в музыке. Правда?</a></li>
                </ul>
            </li>  
            <li><a href="#результаты-исследования">Результаты исследования.</a></li>
        </ol>
    </div>
    <div style="text-align: right; margin-top: 20px; font-style: italic; color: #7f7f7f;">
        Автор: Субботина Алина
    </div>
</div>

<a name="получение-данных"></a>
# Этап 1. Получение данных

Изучу данные, предоставленные сервисом для проекта.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r'C:\Users\alina\dataframes\yandex_music.csv')

In [3]:
df.head(10)

,userID,Track,artist,genre,City,time,Day
0,FFB692EC,Kamigata To Boots,The Mass Missile,rock,Saint-Petersburg,20:28:33,Wednesday
1,55204538,Delayed Because of Accident,Andreas Rönnberg,rock,Moscow,14:07:09,Friday
2,20EC38,Funiculì funiculà,Mario Lanza,pop,Saint-Petersburg,20:58:07,Wednesday
3,A3DD03C9,Dragons in the Sunset,Fire + Ice,folk,Saint-Petersburg,08:37:09,Monday
4,E2DC1FAE,Soul People,Space Echo,dance,Moscow,08:34:34,Monday
5,842029A1,Преданная,IMPERVTOR,rusrap,Saint-Petersburg,13:09:41,Friday
6,4CB90AA5,True,Roman Messer,dance,Moscow,13:00:07,Wednesday
7,F03E1C1F,Feeling This Way,Polina Griffith,dance,Moscow,20:47:49,Wednesday
8,8FA1D3BE,И вновь продолжается бой,NaN,ruspop,Moscow,09:17:40,Friday
9,E772D5C0,Pessimist,NaN,dance,Saint-Petersburg,21:20:49,Wednesday


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63848 non-null  object
 2   artist    57876 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


---
Всего в таблице 7 столбцов:

* userID — идентификатор пользователя;
* Track — название трека;  
* artist — имя исполнителя;
* genre — название жанра;
* City — город, в котором происходило прослушивание;
* time — время, в которое пользователь слушал трек;
* Day — день недели.

Количество значений в столбцах различается. Это говорит о том, что в столбцах с названиями треков, именами исполнителей и названиями жанров есть пропущенные значения.

Нужно решить 2 проблемы: **пропуски** и **некачественные названия столбцов** (присутствуют заглавные буквы и нужно проверить наличие пробелов). Для проверки рабочих гипотез особенно ценны столбцы *`genre`*, *`City`*, *`time`* и *`day`*.

<a name="предобработка-данных"></a>
# Этап 2. Предобработка данных

### <a href="#оглавление">Перейти к оглавлению</a>

In [5]:
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

**Переименую столбцы** для удобства дальнейшей работы и проверю результат.


In [6]:
# new_names = ['user_id', 'track_name', 'artist_name', 'genre_name', 'city', 'time', 'weekday']
# df.set_axis(new_names, axis = 'columns', inplace = True)

In [7]:
df.columns = df.columns.str.lower().str.strip()
df.rename(columns={'userid': 'user_id'}, inplace=True)
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

**Проверю данные на наличие пропусков** вызовом набора методов для суммирования пропущенных значений.

In [8]:
df.isnull().sum()

user_id       0
track      1231
artist     7203
genre      1198
city          0
time          0
day           0
dtype: int64

Пустые значения свидетельствуют, что для некоторых треков доступна не вся информация. Причины могут быть разные: скажем,  не назван конкретный исполнитель народной песни. Хуже, если проблемы с записью данных.

Удаляю в столбце с жанрами пустые значения; убеждаюсь, что их больше не осталось.

In [9]:
df.dropna(subset = ['genre'], inplace = True)

In [10]:
df.isnull().sum()

user_id       0
track       570
artist     6372
genre         0
city          0
time          0
day           0
dtype: int64

Заменяю пропущенные значения в столбцах с названием трека и исполнителя на строку *`'unknown'`*.

In [11]:
df[['track', 'artist']] = df[['track', 'artist']].fillna('unknown')

In [12]:
df.isnull().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

Теперь **необходимо установить наличие дубликатов**.  Если найдутся, удалю их, и проверю, все ли удалились.

In [13]:
df.duplicated().sum()

3755

In [14]:
df = df.drop_duplicates().reset_index(drop = True)

In [15]:
df.duplicated().sum()

0

Дубликаты могли появиться вследствие сбоя в записи данных. Стоит обратить внимание и разобраться с причинами появления такого «информационного мусора».

Выведу уникальные значения столбца с жанрами, чтобы проверить, нет ли среди них неявных дубликатов. 

In [16]:
sorted(df['genre'].unique())

['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'alternativepunk',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'author',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'chanson',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',


In [17]:
df['genre'].nunique()

289

Всего 289 жанров.

Объявлю функцию *`correct_wrong_genre()`*, которая заменяет  неправильное название жанра в столбце *`genre`* на правильное и проверяет успешность выполнения замены.

In [18]:
def correct_wrong_genre(df, wrong, right):
    df['genre'] = df['genre'].replace(wrong, right)
    count_wrong = df[df['genre'] == wrong]['genre'].count()
    print (f"{wrong} исправлен на {right}. Неправильных значений: {count_wrong}")

Исправлю различные варианты названия жанра хип-хоп в таблице (*`hip`*, *`hop`*, *`hip-hop`*) на *`hiphop`*.

In [19]:
correct_wrong_genre(df, 'hip', 'hiphop')
correct_wrong_genre(df, 'hop', 'hiphop')
correct_wrong_genre(df, 'hip-hop', 'hiphop')

hip исправлен на hiphop. Неправильных значений: 0
hop исправлен на hiphop. Неправильных значений: 0
hip-hop исправлен на hiphop. Неправильных значений: 0


Убежусь, что количество уникальных значений жанров уменьшилось.

In [20]:
df['genre'].nunique()

286

Проверю, какие значения есть в столбцах с городом и днем недели.

In [21]:
df['city'].unique()

array(['Saint-Petersburg', 'Moscow'], dtype=object)

In [22]:
df['day'].unique()

array(['Wednesday', 'Friday', 'Monday'], dtype=object)

В исследовании использованы данные по Москве и Санкт‑Петербургу за три дня недели: понедельник, среду и пятницу.

Получаю общую информацию о данных и убеждаюсь, что чистка выполнена успешно.

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60126 entries, 0 to 60125
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  60126 non-null  object
 1   track    60126 non-null  object
 2   artist   60126 non-null  object
 3   genre    60126 non-null  object
 4   city     60126 non-null  object
 5   time     60126 non-null  object
 6   day      60126 non-null  object
dtypes: object(7)
memory usage: 3.2+ MB


---
**Вывод:**

На этапе предобработки в данных обнаружились не только пропуски и проблемы с названиями столбцов, но и всяческие виды дубликатов. Их удаление позволит провести анализ точнее. Поскольку сведения о жанрах важно сохранить для анализа, не просто удаляем все пропущенные значения, а заполним пропущенные имена исполнителей и названия треков. Имена столбцов теперь корректны и удобны для дальнейшей работы.

---

<a name="проверка-гипотез"></a>
# Этап 3. Проверка гипотез

### <a href="#оглавление">Перейти к оглавлению</a>

<a name="гипотеза1"></a>
## Действительно ли музыку в разных городах слушают по-разному?

Была выдвинута гипотеза, что активность пользователей зависит от дня недели, причём в Москве и Петербурге это проявляется по-разному. Проверю это предположение по данным о трёх днях недели — понедельнике, среде и пятнице. Для каждого города устанавлю количество прослушанных  в эти дни композиций с известным жанром и сравнивню результаты.

Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.

In [24]:
df.groupby('city')['genre'].count()

city
Moscow              41892
Saint-Petersburg    18234
Name: genre, dtype: int64

В Москве прослушиваний больше, чем в Питере, но это еще не значит, что Москва более активна. У Яндекс.Музыки может быть больше пользователей в Москве, поэтому проверю, что величины сопоставимы.

In [25]:
df.groupby('city')['user_id'].nunique()

city
Moscow              29043
Saint-Petersburg    12305
Name: user_id, dtype: int64

In [26]:
round(df.groupby('city')['genre'].count()/df.groupby('city')['user_id'].nunique(), 2)

city
Moscow              1.44
Saint-Petersburg    1.48
dtype: float64

В среднем за предоставленный период времени 1 пользователь в Москве послушал 1,44 композиции с известным жанром, а в Петербурге 1,48 композиции.

In [27]:
df.groupby('day')['genre'].count()

day
Friday       21482
Monday       20866
Wednesday    17778
Name: genre, dtype: int64

Понедельник и пятница — время для музыки; по средам пользователи немного больше вовлечены в работу. Но нужно проверить это утверждение для каждого города в отдельности.

Создаю функцию *`number_tracks()`*, которая возвращает количество прослушанных композиций в заданный день недели в заданном городе.

In [28]:
def number_tracks(df, day, city):
    return df[(df['day'] == day) & (df['city'] == city)].shape[0]

Соберу информацию о количестве прослушанных композиций в таблицу

In [29]:
columns = ['city', 'monday', 'wednesday', 'friday']
data = [['Moscow',
         number_tracks(df, 'Monday', 'Moscow'),
         number_tracks(df, 'Wednesday', 'Moscow'),
         number_tracks(df, 'Friday', 'Moscow')],
        ['Saint-Petersburg',
         number_tracks(df, 'Monday', 'Saint-Petersburg'),
         number_tracks(df, 'Wednesday', 'Saint-Petersburg'),
         number_tracks(df, 'Friday', 'Saint-Petersburg')]]
table = pd.DataFrame(data = data, columns = columns)
table

,city,monday,wednesday,friday
0,Moscow,15347,10865,15680
1,Saint-Petersburg,5519,6913,5802


---
**Вывод:**

Результаты показывают, что относительно среды музыку в Петербурге и Москве слушают «зеркально»: в Москве пики приходятся на понедельник и пятницу, а в среду время прослушивания снижается. Тогда как в Санкт-Петербурге среда — день самого большого интереса к музыке, а в понедельник и пятницу он меньше, причём почти одинаково меньше.

---

### <a href="#оглавление">Перейти к оглавлению</a>

<a name="гипотеза2"></a>
## Утро понедельника и вечер пятницы — разная музыка или одна и та же?

Теперь найду ответ на вопрос, какие жанры преобладают в разных городах в понедельник утром и в пятницу вечером. Проверю, отличаются ли жанры в зависимости от города. Кроме того есть предположение, что в понедельник утром пользователи слушают больше бодрящей музыки (например, жанра поп), а вечером пятницы — больше танцевальных (например, электронику).

In [40]:
df.groupby(['city', 'day']).size().unstack(fill_value=0)

day,Friday,Monday,Wednesday
city,,,
Moscow,15680,15347,10865
Saint-Petersburg,5802,5519,6913


In [30]:
moscow_df = df[df['city'] == 'Moscow']

In [31]:
spb_df = df[df['city'] == 'Saint-Petersburg']

Создам функцию *`genre_weekday()`*, которая возвращает список жанров по запрошенному дню недели и времени суток с такого-то часа по такой-то.

In [45]:
def genre_weekday(df, day, time1, time2):
    genre_list = df[(df['day'] == day) & (df['time'] >= time1) & (df['time'] <= time2)]
    genre_list_sorted = genre_list.groupby('genre')['genre'].count().sort_values(ascending = False).head(10)
    return genre_list_sorted

Посмотрю, какое самое раннее и самое позднее время есть в данных.

In [41]:
df['time'].min()

'08:00:00'

In [42]:
df['time'].max()

'22:00:59'

Cравню полученные результаты по таблице для Москвы и Санкт-Петербурга в понедельник утром (с 8 до 11) и в пятницу вечером (с 17 до 23).

In [46]:
genre_weekday(moscow_df, 'Monday', '08:00:00', '11:00:00')

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
classical      157
Name: genre, dtype: int64

In [47]:
genre_weekday(spb_df, 'Monday', '08:00:00', '11:00:00')

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: genre, dtype: int64

Популярные жанры в понедельник утром в Питере и Москве оказались похожи: везде, как и предполагалось, популярен поп. Несмотря на это, есть и различие: в Питере в топ-10 входит джаз, а в Москве жанр world.

In [48]:
genre_weekday(moscow_df, 'Friday', '17:00:00', '23:00:00')

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
alternative    163
classical      163
rusrap         142
Name: genre, dtype: int64

In [49]:
genre_weekday(spb_df, 'Friday', '17:00:00', '23:00:00')

genre
pop            256
rock           216
electronic     216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
Name: genre, dtype: int64

В конце недели ситуация не меняется. Поп-музыка всё так же на первом месте. Но в Питере из топ-10 исчезает ruspop и на 10 месте появляется жанр world.

---
**Вывод:**

Гипотеза не подтвердилась. Жанр поп безусловный лидер в обоих городах, а различия в топ-10 минимальны. Однако различия всё же есть: в Москве чаще слушают русскую популярную музыку, в Петербурге — джаз.

---

### <a href="#оглавление">Перейти к оглавлению</a>

<a name="гипотеза3"></a>
## Москва и Питер — две разные столицы, два разных направления в музыке. Правда?

**Гипотеза:** Питер богат своей рэп-культурой, поэтому это направление там слушают чаще, а Москва — город контрастов, но основная масса пользователей слушает попсу.



In [37]:
moscow_genres = moscow_df.groupby('genre')['genre'].count().sort_values(ascending = False)

In [38]:
moscow_genres.head(10)

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

In [39]:
spb_genres = spb_df.groupby('genre')['genre'].count().sort_values(ascending = False)

In [40]:
spb_genres.head(10)

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

---
**Вывод:**

И в Москве, и в Санкт-Петербурге кроме абсолютно популярного жанра поп есть направление русской популярной музыки. Это значит, что интерес к данному жанру шире. А рэп, вопреки предположению, занимает в обоих городах близкие позиции. Таким образом, предпочтения в двух городах довольно схожи.

---

<a name="результаты-исследования"></a>
# Этап 4. Результаты исследования

### <a href="#оглавление">Перейти к оглавлению</a>

Рабочие гипотезы:

* музыку в двух городах — Москве и Санкт-Петербурге — слушают в разном режиме, активность зависит от дня недели;

* списки десяти самых популярных жанров утром в понедельник и вечером в пятницу имеют характерные отличия;

* население двух городов предпочитает разные музыкальные жанры.

**Общие результаты:**

Москва и Петербург сходятся во вкусах: везде преобладает поп музыка. При этом зависимости предпочтений от дня недели в каждом отдельном городе нет — люди постоянно слушают то, что им нравится. Но между городами в разрезе дней недели наблюдается зеркальность относительно среды: Москва больше слушает в понедельник и пятницу, а Петербург наоборот - больше в среду, но меньше в понедельник и пятницу.

В результате первая гипотеза подтверждена, вторая гипотеза не подтверждена и третья не подтверждена.